// inversify-socket-utils/target/lib/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InversifySocketServer = exports.socketRooms = exports.connectedSocket = exports.socketID = exports.socketQueryParam = exports.socketRequest = exports.socketIO = exports.payload = exports.onDisconnect = exports.onMessage = exports.onConnect = exports.controller = exports.interfaces = exports.TYPE = void 0;
const constants_1 = require("./constants");
Object.defineProperty(exports, "TYPE", { enumerable: true, get: function () { return constants_1.TYPE; } });
const decorators_1 = require("./decorators");
Object.defineProperty(exports, "connectedSocket", { enumerable: true, get: function () { return decorators_1.connectedSocket; } });
Object.defineProperty(exports, "controller", { enumerable: true, get: function () { return decorators_1.controller; } });
Object.defineProperty(exports, "onConnect", { enumerable: true, get: function () { return decorators_1.onConnect; } });
Object.defineProperty(exports, "onDisconnect", { enumerable: true, get: function () { return decorators_1.onDisconnect; } });
Object.defineProperty(exports, "onMessage", { enumerable: true, get: function () { return decorators_1.onMessage; } });
Object.defineProperty(exports, "payload", { enumerable: true, get: function () { return decorators_1.payload; } });
Object.defineProperty(exports, "socketID", { enumerable: true, get: function () { return decorators_1.socketID; } });
Object.defineProperty(exports, "socketIO", { enumerable: true, get: function () { return decorators_1.socketIO; } });
Object.defineProperty(exports, "socketQueryParam", { enumerable: true, get: function () { return decorators_1.socketQueryParam; } });
Object.defineProperty(exports, "socketRequest", { enumerable: true, get: function () { return decorators_1.socketRequest; } });
Object.defineProperty(exports, "socketRooms", { enumerable: true, get: function () { return decorators_1.socketRooms; } });
const interfaces = __importStar(require("./interfaces"));
exports.interfaces = interfaces;
const server_1 = require("./server");
Object.defineProperty(exports, "InversifySocketServer", { enumerable: true, get: function () { return server_1.InversifySocketServer; } });
